src/ramdb.cpp

```diff
@@ -0,0 +1,390 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/ramdb.hpp>
+
+#include <miopen/errors.hpp>
+#include <miopen/lock_file.hpp>
+#include <miopen/logger.hpp>
+#include <miopen/md5.hpp>
+
+#include <boost/filesystem/operations.hpp>
+#include <boost/filesystem.hpp>
+
+#include <chrono>
+#include <ctime>
+#include <fstream>
+#include <limits>
+#include <map>
+#include <mutex>
+#include <sstream>
+
+namespace miopen {
+
+std::string RamDb::GetTimeFilePath(const std::string& path) { return path + ".time"; }
+
+static ramdb_clock::time_point GetDbModificationTime(const std::string& path)
+{
+    const auto time_file_path = RamDb::GetTimeFilePath(path);
+    auto file                 = std::ifstream{time_file_path};
+
+    if(!file)
+        // Zero time from epoch, considering file ancient if time can't be retrieved.
+        return {};
+
+    ramdb_clock::rep time;
+    file >> time;
+    return ramdb_clock::time_point{ramdb_clock::duration{time}};
+}
+
+static void UpdateDbModificationTime(const std::string& path)
+{
+    MIOPEN_LOG_I2("Updating db modification time for " << path);
+
+    const auto time           = ramdb_clock::now().time_since_epoch();
+    const auto time_file_path = RamDb::GetTimeFilePath(path);
+    auto file                 = std::ofstream{time_file_path};
+
+    if(!file)
+    {
+        MIOPEN_LOG_E("Cannot update database modification time: " + time_file_path);
+        return;
+    }
+
+    file << time.count();
+}
+
+#define MIOPEN_VALIDATE_LOCK(lock)                       \
+    do                                                   \
+    {                                                    \
+        if(!(lock))                                      \
+            MIOPEN_THROW("Db lock has failed to lock."); \
+    } while(false)
+
+static std::chrono::seconds GetLockTimeout() { return std::chrono::seconds{60}; }
+
+using exclusive_lock = std::unique_lock<LockFile>;
+
+RamDb::RamDb(std::string path, bool is_system) : PlainTextDb(path, is_system)
+{
+    if(is_system)
+        MIOPEN_THROW("PlainTextDb class is not supported as system database. Use the ReadOnlyRamDb "
+                     "class instead.");
+}
+
+RamDb& RamDb::GetCached(const std::string& path, bool warn_if_unreadable)
+{
+    static std::mutex mutex;
+    static const std::lock_guard<std::mutex> lock{mutex};
+
+    static bool saved      = false;
+    static auto saved_path = path;
+    static RamDb instance{path, warn_if_unreadable};
+
+    if(saved)
+    {
+        if(saved_path != path)
```

Comment:
This means MIOpen would not support a system with two different GPUs. Can't we just return the instance for that GPU instead ?

Reply:
We currently don't support running on two different types of GPU. We had an ancient discussion with Artem about this. I can change it to a map if we have a plan to support multiple different GPUs in any foreseeable future.

Reply:
Yes, please see https://github.com/AMDComputeLibraries/MLOpen/pull/1968#discussion_r319216292.

Reply:
We have a dedicated issues about that, https://github.com/AMDComputeLibraries/MLOpen/issues/1312